File: calcs.cpp
/* Calculates ACN and space writhe of a piecewise-linear path.
   This is much faster than doing it in python, since python is interpreted and C is compiled
   This formula, as well as all the variable names, is taken from:
   https://en.wikipedia.org/wiki/Writhe#Numerically_approximating_the_Gauss_integral_for_writhe_of_a_curve_in_space

*/
#define PI 3.14159265358979323846

#include <tuple>
#include <math.h>
#include <vector>
#include <iostream>
#include <fstream>
#include <sstream>
#include <thread>

std::vector<std::array<double, 3>> points_csv(std::string filename) {
  // returns contents of file as a list of 3D points
  std::ifstream file (filename);
  std::vector<std::array<double, 3>> points;
  std::string line, coord;
  while (std::getline(file, line)) {
    std::array<double, 3> new_point;
    std::stringstream s (line);
    for (int i = 0; i <3 ; i++) {
      getline(s, coord, ',');
      new_point[i] = std::stod(coord);
    }
    points.push_back(new_point);
  }
  file.close();
  return points;
}

inline double magnitude(std::array<double, 3> vec) {
  return sqrt( vec[0]*vec[0] + vec[1]*vec[1] + vec[2]*vec[2] );
}

inline std::array<double, 3> unit(std::array<double, 3> vec) {
  // returns a vector (math vector, called array in C++) that points
  // in the same direction as the input, but with magnitude = 1
  if (magnitude(vec) == 0.) {
    return {0., 0., 0.};
  }
  return { vec[0] / magnitude(vec), vec[1] / magnitude(vec), vec[2] / magnitude(vec) };
}

inline std::array<double, 3> cross(std::array<double, 3> vec1,
                                   std::array<double, 3> vec2) {
  // returns the cross product of inputs vec1 & vec2
  double x = vec1[1] * vec2[2] - vec1[2] * vec2[1];
  double y = vec1[2] * vec2[0] - vec1[0] * vec2[2];
  double z = vec1[0] * vec2[1] - vec1[1] * vec2[0];
  return { x, y, z };
}

inline double dot(std::array<double, 3> vec1,
                  std::array<double, 3> vec2) {
  // Returns the dot product of inputs vec1 & vec2
  return vec1[0] * vec2[0] + vec1[1] * vec2[1] + vec1[2] * vec2[2];
}

std::array<double, 3> operator - (std::array<double, 3> vec1,
                                  std::array<double, 3> vec2) {
  // Returns vec1 minus vec2
  return { vec1[0]-vec2[0], vec1[1]-vec2[1], vec1[2]-vec2[2] };
}

inline double sign(double num) { return num > 0. ? 1. : num < 0. ? -1. : 0.; }

inline double vec_sum(std::vector<double> vec) {
  double sum = 0;
  for (double num: vec) {
    sum += num;
  }
  return sum;
}

inline double expected_crossing_value(std::array<double, 3> p1,
                                      std::array<double, 3> p2,
                                      std::array<double, 3> p3,
                                      std::array<double, 3> p4) {
  // Returns the chance that line segment p1 -> p2 will cross p3 -> p4
  // when viewed from a random 2D projection. If the crossing is negative,
  // returns the same thing but negative
  std::array<double, 3> n1 = unit(cross(p3 - p1, p4 - p1));
  std::array<double, 3> n2 = unit(cross(p4 - p1, p4 - p2));
  std::array<double, 3> n3 = unit(cross(p4 - p2, p3 - p2));
  std::array<double, 3> n4 = unit(cross(p3 - p2, p3 - p1));
  for (std::array<double, 3> unit_vector : {n1, n2, n3, n4}) {
    // If any of the cross products resulted in magnitude zero,
    // the final result should be zero, and we need to catch that now
    if (magnitude(unit_vector) == 0.) {
      return 0.;
    }
  }
  double omega_star = asin(dot(n1, n2)) +
                      asin(dot(n2, n3)) +
                      asin(dot(n3, n4)) +
                      asin(dot(n4, n1));
  omega_star /= 4 * PI;
  double omega_signed = omega_star * sign(dot(p3 - p1, cross(p4 - p3, p2 - p1)));
  return omega_signed;
}

void add_crossing_values(int i,
                         std::vector<std::array<double, 3>>* knot_1,
                         std::vector<std::array<double, 3>>* knot_2,
                         std::vector<double>* acn_subcalcs,
                         std::vector<double>* space_writhe_subcalcs) {
  // Compares segment i of knot_1 to all segments of knot_2
  // This is to break up the calculations. This can be run in one thread,
  // but it would be slow to compare every segment of knot_1 to every segment
  // of knot_2 in just one thread. Since we don't want multiple threads touching
  // the same final calcs, this modifies one element of a vector passed by
  // pointer. That's the vector of all the subcalcs, which will be added together later.
  for (int j = 1; j < knot_2->size(); j++) {
    if (i == j) { continue; }
    std::array<double, 3> p1 = knot_1->at(i - 1);
    std::array<double, 3> p2 = knot_1->at(i);
    std::array<double, 3> p3 = knot_2->at(j - 1);
    std::array<double, 3> p4 = knot_2->at(j);
    double crossing_value = expected_crossing_value(p1, p2, p3, p4);
    space_writhe_subcalcs->at(i) += crossing_value;
    acn_subcalcs->at(i) += fabs(crossing_value);
  }
}

std::tuple<double, double> acn_and_writhe(std::vector<std::array<double, 3>> knot) {
  // Given a knot as a vector of points, returns the tuple (ACN, space writhe)
  // Since the calculation is a bunch of smaller things to be added together,
  // it splits that up into knot.size() different groups of calcs (subcalcs)
  // which each have their own thread. They modify values in an vector of subcalcs,
  // which is then added up to create the final calculations
  std::vector<double> acn_subcalcs (knot.size());
  std::vector<double> space_writhe_subcalcs (knot.size());
  std::vector<std::thread> threads = {};
  for (int i = 1; i < knot.size(); i++) {
    std::thread new_thread(add_crossing_values, i, &knot, &knot, &acn_subcalcs, &space_writhe_subcalcs);
    threads.push_back(std::move(new_thread));
  }
  for (std::thread& current_thread: threads) {
    current_thread.join();
  }
  double acn = vec_sum(acn_subcalcs);
  double space_writhe = vec_sum(space_writhe_subcalcs);
  return std::make_tuple(acn, space_writhe);
}

std::tuple<double, double> ACN_and_linking_number(
    std::vector<std::array<double, 3>> knot_A,
    std::vector<std::array<double, 3>> knot_B) {
  // This is very similar to the acn and space writhe calculations (see above)
  // except that it's iterating over two different knots. That is, it's
  // comparing each segment of one knot to a segment in the other, instead of
  // comparing every segment in a knot to every other segment in the same knot
  std::vector<double> acn_subcalcs (knot_A.size());
  std::vector<double> space_writhe_subcalcs (knot_A.size());
  std::vector<std::thread> threads = {};
  for (int i = 1; i < knot_A.size(); i++) {
    std::thread new_thread(add_crossing_values, i, &knot_A, &knot_B, &acn_subcalcs, &space_writhe_subcalcs);
    threads.push_back(std::move(new_thread));
  }
  for (std::thread& current_thread: threads) {
    current_thread.join();
  }
  // In this context, ACN is the number of times one knot crosses the other
  double acn = vec_sum(acn_subcalcs);
  double linking_number = vec_sum(space_writhe_subcalcs);
  return std::make_tuple(acn, linking_number);
}

int main() {
  std::vector<std::array<double, 3>> points_A = points_csv("points_A.csv");
  std::vector<std::array<double, 3>> points_B = points_csv("points_B.csv");
  double lnk_num = std::get<1>(ACN_and_linking_number(points_A, points_B));
  std::cout << "Linking number: " << lnk_num << std::endl;
  std::tuple<double, double> tuple_acn_and_space_writhe = acn_and_writhe(points_A);
  double acn          = std::get<0>(tuple_acn_and_space_writhe);
  double space_writhe = std::get<1>(tuple_acn_and_space_writhe);
  std::cout << "Hardware concurrency: " << std::thread::hardware_concurrency() << std::endl;
  std::cout << "ACN: " << acn << "\nSpace Writhe: " << space_writhe << std::endl;
  return 0;  // executed successfully
}

File: calcs.py
"""
A collection of functions to use in analyzing the SARS-CoV2 spike protein

Throughout this file, groups of points will be stored mostly as matrices with
each row representing a point, and each column representing a dimension (x, y, or z)
"""
import math
import numpy as np
from io import StringIO

def backbone(protein_pdb: StringIO, chain="all chains", CA_only=True):
    # Uses a PDB file to get the coordinates of atoms in the backbone of a protein chain
    pdb_data = protein_pdb.read()
    atoms = []
    backbone_atoms = ['CA'] if CA_only else ['CA', 'C', 'N']
    for line in pdb_data.split('\n'):
        # By default, look at all chains, unless just one is specified
        if line[0:4] == 'ATOM' and (chain is "all chains" or line[21] == chain) and line[13:16].rstrip(' ') in backbone_atoms:
            residue_num = int(line[24:26])
            x = float(line[31:38].strip(' '))
            y = float(line[39:46].strip(' '))
            z = float(line[47:54].strip(' '))
            atoms.append([x, y, z])
    return np.matrix(atoms)

def dump_beads(*components, file_name='beads.txt') -> None:
    # Given a list of atom coordinates, creates a file for knotplot to open
    # Find centroid of all chains, in order to center it
    centroid = np.array([[0., 0., 0.]])
    total_atoms = 0
    for component in components:
        for atom in component:
            centroid += atom[0]
            total_atoms += 1
    centroid /= total_atoms
    for component in components:
        for atom in component:
            atom -= centroid
    knotplot_data = ''
    components = [component.tolist() for component in components]
    for component_num in range(len(components)):
        knotplot_data += f'Component {component_num + 1} of {len(components)}:\n'
        component = components[component_num]
        for atom in component:
            knotplot_data += ' '.join([str(coord) for coord in atom]) + '\n'
    with open(file_name, 'w+') as knotplot_file:
        knotplot_file.write(knotplot_data)

def rmsd(chain1, chain2):
    # TODO: create functions for just getting the centroid or rotation matrix
    n = len(chain1)
    if n != len(chain2): raise Error
    chain1 = np.matrix(chain1)
    chain2 = np.matrix(chain2)
    # Center each group of points by subtracting their respective centroids
    chain1 = chain1 - np.ones((n, n)) * chain1 / n
    chain2 = chain2 - np.ones((n, n)) * chain2 / n
    # Use Kabsch algorithm to line up chain1 and chain2 in order to minimize RMSD
    # The Kabsch algo finds the rotation matrix that best maps chain1 onto chain2
    # https://en.wikipedia.org/wiki/Kabsch_algorithm
    covariance_matrix = chain1.T * chain2
    U, S, VT = np.linalg.svd(covariance_matrix)
    d = np.linalg.det(U * VT)
    rotation_matrix = U * np.diag([1,1,d]) * VT
    chain1 *= rotation_matrix
    total_square_diff = sum([row * row.T for row in chain1 - chain2])[0, 0]
    return math.sqrt(total_square_diff / n)

def unit_vec(vector):
    if np.linalg.norm(vector) == 0:
        raise ZeroDivisionError
    return vector / np.linalg.norm(vector)

def sign(num): return 1 if num > 0 else -1 if num < 0 else 0

def ACN_and_space_writhe(chain) -> (float, float):
    # calculates ACN & space writhe of a piecewise-linear path, which is given as a matrix of points
    # returns ACN and space writhe together, as a tuple of 2 floats
    # This formula, as well as all the variable names, is taken from:
    # https://en.wikipedia.org/wiki/Writhe#Numerically_approximating_the_Gauss_integral_for_writhe_of_a_curve_in_space
    acn = 0
    space_writhe = 0
    # Iterate through every pair of line segments
    for i in range(1, len(chain)):
        for j in range(1, i - 1):
            # p1, p2, p3, & p4 are endpoints of line segments i & j
            p1 = chain[i - 1]
            p2 = chain[i]
            p3 = chain[j - 1]
            p4 = chain[j]
            n1 = unit_vec(np.cross(p3 - p1, p4 - p1))
            n2 = unit_vec(np.cross(p4 - p1, p4 - p2))
            n3 = unit_vec(np.cross(p4 - p2, p3 - p2))
            n4 = unit_vec(np.cross(p3 - p2, p3 - p1))
            omega_star = math.asin(n1.dot(n2.T)[0, 0]) + \
                         math.asin(n2.dot(n3.T)[0, 0]) + \
                         math.asin(n3.dot(n4.T)[0, 0]) + \
                         math.asin(n4.dot(n1.T)[0, 0])
            omega_star /= 2 * math.pi
            omega_signed = omega_star * sign(np.cross(p4 - p3, p2 - p1).dot((p3 - p1).T)[0, 0])
            acn += omega_star
            space_writhe += omega_signed
    return (acn, space_writhe)

def rog(chain) -> float:
    # given a cluster of points (as an n by 3 matrix), calculates the radius of gyration
    # For now, assumes all points have the same weight
    moment = sum(chain)
    centroid = moment / len(chain)
    sum_of_distances_squared = sum([pow(np.linalg.norm(point - centroid), 2) for point in chain])
    return math.sqrt(sum_of_distances_squared / len(chain))
